shenyu-proxy/proxy-remote/src/main/java/org/apache/shenyu/proxy/remote/netty/NettyChannel.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shenyu.proxy.remote.netty;
+
+import io.netty.channel.ChannelFutureListener;
+import org.apache.shenyu.proxy.remote.Channel;
+import org.apache.shenyu.proxy.remote.ChannelFuture;
+
+import java.net.SocketAddress;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * NettyChannel .
+ */
+public class NettyChannel implements Channel {
+    
+    private static final ConcurrentMap<io.netty.channel.Channel, NettyChannel> CHANNEL_CACHE = new ConcurrentHashMap<>();
+    
+    private final io.netty.channel.Channel channel;
+    
+    /**
+     * Instantiates a new Netty channel.
+     *
+     * @param channel the channel
+     */
+    public NettyChannel(final io.netty.channel.Channel channel) {
+        this.channel = channel;
+    }
+    
+    /**
+     * send.
+     * Send data to network channel.
+     *
+     * @param message the message
+     * @return the channel future
+     */
+    @Override
+    public ChannelFuture send(final Object message) {
+        io.netty.channel.ChannelFuture channelFuture = channel.writeAndFlush(message);
+        return new NettyChannelFuture(channelFuture);
+    }
+    
+    /**
+     * Determine if it is connected.
+     *
+     * @return boolean boolean
+     */
+    @Override
+    public boolean isConnected() {
+        return channel.isActive();
+    }
+    
+    /**
+     * remote address.
+     *
+     * @return socket address
+     */
+    @Override
+    public SocketAddress remoteAddress() {
+        return channel.remoteAddress();
+    }
+    
+    /**
+     * Local address socket address.
+     *
+     * @return the socket address
+     */
+    @Override
+    public SocketAddress localAddress() {
+        return channel.localAddress();
+    }
+    
+    /**
+     * Is opened boolean.
+     *
+     * @return the boolean
+     */
+    @Override
+    public boolean isOpened() {
+        return channel.isOpen();
+    }
+    
+    /**
+     * Is close boolean.
+     *
+     * @return the boolean
+     */
+    @Override
+    public boolean isClose() {
+        return !this.isOpened();
+    }
+    
+    /**
+     * Close.
+     */
+    @Override
+    public void close() {
+        channel.close().addListener((ChannelFutureListener) channelFuture -> {
+            removeChannel(channel);
+        });
+    }
+    
+    /**
+     * Get a channel, create a Netty Channel if it doesn't exist.
+     *
+     * @param channel the channel nettyChannel.
+     * @return the or add channel
+     */
+    public static NettyChannel getOrAddChannel(final io.netty.channel.Channel channel) {
+        NettyChannel nettyChannel = CHANNEL_CACHE.get(channel);
+        
+        if (nettyChannel == null) {
+            NettyChannel ret = new NettyChannel(channel);
+            if (channel.isActive()) {
```

Comment:
no need to judge, as following code set the value if nettyChannel is null again

Reply:
The first time you create a channel, it might be empty.